subprojects/packagefiles/openssl/generator.sh

```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
+set -e
+
+cd "$(dirname "${BASH_SOURCE[0]}")"
+
+# Node.js version should bundle OpenSSL of matching version to one specified in wrap file
+node_version=v16.5.0
+openssl_version=$(grep 'directory = ' ../../openssl.wrap | grep -oE '[0-9]+\.[0-9]+\.[0-9]+[a-z]')
+
+rm -rf node
+git clone --depth 1 --branch $node_version https://github.com/nodejs/node.git
+
+rm -rf generated-config
+
+pushd node/deps/openssl
+
+# Apply patch that will allow us generate `meson.build` for different targets
+patch -u config/generate_gypi.pl -i ../../../generate_gypi.pl.patch
```

Comment:
Instead of patching a third party perl script to generate our meson.build, would it be possible to write our own python script for that? Unless that would be too complicated?

Reply:
The file is https://github.com/nodejs/node/blob/master/deps/openssl/config/generate_gypi.pl

It should be possible to rewrite in Python, but we would also need to swap templating language at the same time. I'm not looking forward to that, but if someone is adventurous, it might be worth pursuing.

Eventually we would probably want to get rid of Node.js completely as there will be not much left after that script, we can just copy a few more files and make it completely standalone.

Reply:
I think the point is that this is deep inside the entire business logic of the perl build system, and literally runs the openssl `./Configure` to generate basically everything.

Reply:
If we can make it standalone, even by copying a few perl scripts, that would be great.

My concern here is we cannot review generated code, we can review the script that generates it. So the more we can simplify the code path, the easier it is to convince we build a legitimate OpenSSL.

Reply:
There is a small makefile and 170 lines of perl, is that all that is needed from that node.js repo? If so, I think it's worth porting to Python and make it standalone in wrapdb. Or am I missing something?

Reply:
I've been looking a bit at what the makefile/perl does, and it seems doable in our own python script. You can import configdata.pm into python like that:
```
stdout = subprocess.check_output(['perl', '-MJSON', '-I.', '-Mconfigdata', '-e', 'print encode_json(\%unified_info);'])
unified_info = json.loads(stdout)
```

Then it's just a matter to create the `generated_srcs` list and call `make`.

Just trying to make this readable, I hate perl so much...

Reply:
I believe there are a few header files and other minor things, but in general it seems doeable.

Reply:
So, I played a bit more to port generator to a standalone python script: https://github.com/xclaesse/openssl/commit/57406a086854d5e359044780c78393230fcd4e35. It's not complete, but I think it's totally doable and would be less hackish than depending on node.js

Reply:
Updated my python script to also generate parts of meson.build: https://github.com/xclaesse/openssl/commit/1c1ae196bbd9271410a4ec84f69b9f9b04383945.

Reply:
Almost there, just a few bits missing in generated meson.build files, but it now contains all needed generated files: https://github.com/xclaesse/openssl/commit/33f90a032c55697bc5a4be4896437d386cb7c0dc.

Reply:
Hmm, so there was a catch... Generating asm for Windows requires nmake instead of GNU make, and that does not seems to be available on Linux. Node.js tackly that by handwriting 5k likes of Makefile and run it on Linux... I'm not fan of that solution at all...

The good news is I updated my script (https://github.com/xclaesse/openssl/commit/67aba4465f8c6e7044c869f2d2c433a71e789adb) to work on Windows, and I think `generated-config` is now 1:1 identical to node.js.

What about we add Windows CI to wrapdb (we should do that anyway) and somewhat get the CI combine generated code from both linux and windows runs? That's something I know how to do with gitlab artifacts, but I assume github is also capable of having jobs that depend on each other and share artifacts, no?

Reply:
That sounds like a very complicated process for one-off use case of OpenSSL.

Another (weirder) alternative is to take advantage of the fact that on Windows you can run both Windows and Linux containers with Docker, thus CI can run on Windows, run part of the actions under Linux container and the other part under Windows container and combine the artifacts together at the end.

With all that in mind using makefiles maintained by Node.js folks doesn't sound that bad of an idea after all :upside_down_face:

Reply:
I was hoping to run nmake under wine, but seems Microsoft does not ship standalone nmake.exe. Does github allow CI workflow under Windows Subsystem for Linux? AFAIK that's the only env where we can run both nmake and gnu make :P

I think doing a Windows CI is not that bad, because we should test our wraps on Windows too anyway. 2 birds one stone...

But in the meantime I guess we should just download node.js makefiles?

Reply:
I would still say we should merge it without assembly first and then tackle assembly next ;)

Reply:
I think it makes sense to merge non-Python version that already works first and then migrate to Python when all pieces for that are in place. Also Windows isn't using assembly in the current state of the PR anyway.

Reply:
I support whatever is the fastest route that doesn't involve checking in tons of unreviewable assembly files :smile:

Reply:
Current PR revision only includes assembly files in the generated release patch, but not in the repo. Files are generated as part of release process in CI (see updates to `tools/create_release.py`).

Here is an example in my fork of what it looks like: https://github.com/nazar-pc/wrapdb/releases/tag/openssl_1.1.1k-1